ReactCommon/jsi/JSCRuntime.cpp

```diff
@@ -286,9 +286,11 @@ class JSCRuntime : public jsi::Runtime {
 namespace {
 std::string JSStringToSTLString(JSStringRef str) {
   size_t maxBytes = JSStringGetMaximumUTF8CStringSize(str);
-  std::vector<char> buffer(maxBytes);
-  JSStringGetUTF8CString(str, buffer.data(), maxBytes);
-  return std::string(buffer.data());
+  char* buffer = new char[maxBytes];
```

Comment:
That's very strange. I don't really understand *why* the previous implementation is slower.
When I read the changed code without looking at the previous version of it, I thought "oh, using std::vector is more ideomatic here and should be as fast as a plain buffer".

I think the most efficient way to implement this function is to use something like `folly::small_vector` (or simply use the same approch manually) trying to avoid heap allocation for small strings.

Reply:
Oh my, look at this:
https://en.cppreference.com/w/cpp/container/vector/vector
and then look at this `3-4) Linear in count`. That's probably because we need to set to `0` every single element.

Reply:
@shergin AFAICT it's not legal to use `vector::data` after `reserve`, and actually resizing the vector runs into the same needless zero-initialisation issue. So I think there's no way to do this with `vector` without paying this cost.

Reply:
@motiz88 is right, I am wrong. The unque_ptr looks like the best option.

Reply:
(As a huge a huge fan of SSO, I would suggest to implement it here as well (e.g. having if the size is smaller than 21 or whatever, use stack-allocated buffer), but that's out of the scope of this PR.)